### ISSUE TITLE
Function signature confusion after ternary operator

### ISSUE THREAD

**Issue by Mickael-van-der-Beek** [microsoft/TypeScript#35707]
I'm expecting this to not be a bug but I can't seem to figure out what else it could be.
Basically I'm trying to create a `(TLS)Socket` instance using the core Node.js `net` or `tls` modules depending on a condition.
Instantiating them by directly referencing the module works fine but whenever I introduce a level of indirection by conditionally defining the TCP client module, compilation breaks.

**Code**

```ts
import net from 'net';
import tls from 'tls';

const socketA = net.connect({ host: 'localhost', port: 1234 });
const socketB = tls.connect({ host: 'localhost', port: 1234 });

const protocol = process.argv[2];
const client = protocol === 'https' ? tls : net;
const socketC: net.Socket | tls.TLSSocket = client.connect({ host: 'localhost', port: 1234 });

console.log(socketA, socketB, socketC);
```

**Expected behavior:**

Successful compilation.

**Actual behavior:**

```
This expression is not callable.
  Each member of the union type
    '{ (options: NetConnectOpts, connectionListener?: (() => void) | undefined): Socket;
     (port: number, host?: string | undefined, connectionListener?: (() => void) | undefined): Socket; 
     (path: string, connectionListener?: (() => void) | undefined): Socket; } | { ...; }'
has signatures, but none of those signatures are compatible with each other.

12 const socketC = client.connect({ host: 'localhost', port: 443 });
                          ~~~~~~~
```

**Playground Link:**

[Playground Link](https://www.typescriptlang.org/play/?module=1#code/OQVwzgpgBGAuBOBLAxrYBuAUJxBbADgPbyxQB2EpAZvIblMBWlnkSVLADZhQ10NcwGbMkJk4MQsgDWlAIJQAvOUoA6UWQqoAFAG8oAC0JwAXA05SAhpyNxgAGihtYZgIwAmAMwAWKAF8ASiwNCTApWVgAISUObnUxLVg9Q2MXcysbVIcnYjSPH38gkTEJfFpYKUJOGLKpCDAwVUt4AHMANwBtdwBdYJLSZE5ECDJSZVqK0WrFGYYDWFh8ISgAflieMyY+8VIwmUoAYU21AGVwyigAHwACwVUAFQAZE7P9sahB4dH4zQgdfVsaWAFmQ1kB2Wcbi8vkCWEwISqEFUFha2j2ETkjnRlEiWPOsAOQSAA)

**Related Issues:**

**Comment by jcalz** [comment]
This is essentially a duplicate of #7294 (this part of which has not been fixed by #29011).  If you have a union of function types and more than one of the members of the union is overloaded or generic, the compiler won't let you call it.  You need to help the compiler out by manually annotating the union as a single callable signature it can verify, e.g., 

```ts
const client: {
    connect(opts: tls.ConnectionOptions & net.ConnectOpts): net.Socket | tls.TLSSocket
} = protocol === 'https' ? tls : net; // no error

const socketC = client.connect({ host: 'localhost', port: 1234 }); // okay
```
[Playground Link](https://www.typescriptlang.org/play/?module=1#code/OQVwzgpgBGAuBOBLAxrYBuAUJxBbADgPbyxQB2EpAZvIblMBWlnkSVLADZhQ10NcwGKNmSEycGIWQBrSgEEoAXnKUAdGLIVUACgDeUABaE4ALgadpAQ07G4wADRQ2scwEYATAGYALFAC+AJRYmpJg0nKwAELKHNwa4tqw+kYmrhbWtmmOzsTpnr4BwaLikvi0sNKEnLHl0hBgYGpW8ADmAG4A2h4AuiGlpMiciBBk6XqYUFNQmkk6hPiwYOaCagDCiRCoiOIA8os7ElAAZKqw65uo+0uB5kxqAMoRlFAAPnFNACoAMg9PspRMP5ahUqjUlBCGIZYLB8EIoAB+D5QO6UdBQAD0GPIhCgEHgtHgJQkpHCANga1iQxGYwSWi2yQMdnSwEsyBszJyLnc3j8QXRWKghBkVgAnsTwpwIGpLK0dGTIvInArKFFlc8KcEgA)

**Comment by Mickael-van-der-Beek** [comment]
Thank you for the quick and complete answer. I'll keep an eye on the related issues as well. This can be closed.
